ramips: add support for 7Links WLR-1230 and WLR-1240

s-2 opened the pull request:
The WLR-1230 (ZX-5207) is an AC1200 Wave 2 outdoor repeater
with sector antennas for wall or pole mounting.

The WLR-1240 (ZX-5434) is an AC1200 Wave 2 outdoor repeater
with omnidirectional antennas for wall or pole mounting.

The devices are manufactured by Todaair and meant to be used with
a tuya-based app, there is no webinterface for configuration.

Specifications:
- MT7628AN, 8 MiB SPI NOR flash, 64 MiB RAM, 2x2 802.11n
- MT7613 2x2 802.11ac Wave 2
- 802.3af PoE or 12V 1A 5.5x2.1 power supply (included)
WLR-1230:
- 3 LEDs WLAN, LAN, RES; PWR LED is not software-controllable
WLR-1240
- top RGB LED ring

TFTP installation:
- rename sysupgrade to `firmware_auto.bin`
- provide at 192.168.1.10 during boot

HTTP installation:
- keep reset button pressed for 5 seconds during power on (LEDs
  flash slowly, then quickly to confirm, then remain steady on)
- recovery web interface is at 192.168.1.1, upload sysupgrade

Opening WLR-1230:
- two screws are located in the bottom left and right corners
  underneath the label, inner tray slides out easily

Opening WLR-1240:
- use suction cup to remove top cap within LED ring
- two screws are located in holes underneath silicone sealant
- two further screws are located at the bottom
  
initramfs boot
- open device, connect serial console (pins are labelled)
- keep pressing `4` during second tftp attempt to enter uboot shell
- run `tftpboot 82000000` to avoid memory overlap, then `bootm`


--------------------------------

The LAN LED on WLR-1230 only shows link state, but does not flash on activity.
Is this intentional or did I miss something in the configuration?

For both devices, LuCI would complain `switch0 has an unknown topology` and all ports are shown,
although only port 4 is exposed. I'm not sure how to avoid this using dts (`portmap` / `portdisable`) or `02_network`.

--------------------------------

The devices are imported to Germany by Pearl, order numbers and current prices are:

WLR-1230 (flat design, sector antenna)
- ZX-5207 single pack (new) € 69,99
- ZX-5225 double pack (new) € 129,99
- REF-52078 single pack (refurbished) € 56,99

WLR-1240 (cylindric design, omnidirectional antenna)
- ZX-5434 single pack (new) € 58,99
- ZX-5435 double pack (new)	€ 115,99
- REF-54348 single pack (refurbished) € 51,91

They might be interesting for use with gluon, I will prepare and test a backport to 23.05.

s-2 committed:
ramips: add support for 7Links WLR-1230

The WLR-1230 (ZX-5207) is an AC1200 Wave 2 outdoor repeater
with sector antennas for wall or pole mounting.
The device is manufactured by Todaair and meant to be used with
a tuya-based app, there is no webinterface for configuration.

Specifications:
- MT7628AN, 8 MiB SPI NOR flash, 64 MiB RAM, 2x2 802.11n
- MT7613 2x2 802.11ac Wave 2
- 802.3af PoE or 12V 1A 5.5x2.1 power supply (included)
- 3 LEDs WLAN, LAN, RES; PWR LED is not software-controllable

TFTP installation:
- rename sysupgrade to `firmware_auto.bin`
- provide at 192.168.1.10 during boot

HTTP installation:
- keep reset button pressed for 5 seconds during power on (LEDs
  flash slowly, then quickly to confirm, then remain steady on)
- recovery web interface is at 192.168.1.1, upload sysupgrade

Opening the device
- two screws are located in the bottom left and right corners
  underneath the label, inner tray slides out easily

initramfs boot
- open device, connect serial console (pins are labelled)
- keep pressing `4` during second tftp attempt to enter uboot shell
- run `tftpboot 82000000` to avoid memory overlap, then `bootm`

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 committed:
ramips: add support for 7Links WLR-1240

The WLR-1240 (ZX-5434) is an AC1200 Wave 2 outdoor repeater
with omnidirectional antennas for wall or pole mounting.
The device is manufactured by Todaair and meant to be used with
a tuya-based app, there is no webinterface for configuration.

Specifications:
- MT7628AN, 8 MiB SPI NOR flash, 64 MiB RAM, 2x2 802.11n
- MT7613 2x2 802.11ac Wave 2
- 802.3af PoE or 12V 1A 5.5x2.1 power supply (included)
- top RGB LED ring

TFTP installation:
- rename sysupgrade to `firmware_auto.bin`
- provide at 192.168.1.10 during boot

HTTP installation:
- keep reset button pressed for 5 seconds during power on (light blue
  LED flashes slowly, then quickly to confirm, then remains steady on)
- recovery web interface is at 192.168.1.1, upload sysupgrade

Opening the device
- use suction cup to remove top cap within LED ring
- two screws are located in holes underneath silicone sealant
- two further screws are located at the bottom

initramfs boot
- open device, connect serial console (pins are labelled)
- keep pressing `4` during second tftp attempt to enter uboot shell
- run `tftpboot 82000000` to avoid memory overlap, then `bootm`

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

DragonBluep commented on target/linux/ramips/dts/mt7628an_7links_wlr-12xx.dtsi:
```diff
@@ -0,0 +1,129 @@
+					};
+
+					macaddr_factory_4: macaddr@44 {
+						compatible = "mac-address";
```
Does any linux driver support it? If not, please remove it.

s-2 replied:
Indeed, only few other devices have `mac-base` here, but the cell name is usually just given when referencing.
Will remove.

DragonBluep commented on target/linux/ramips/dts/mt7628an_7links_wlr-12xx.dtsi:
```diff
@@ -0,0 +1,129 @@
+					};
+
+					eeprom_factory_8000: eeprom@8000 {
+						reg = <0x8000 0x200>;
```
MT7663 eeprom size is 0x4da8.

s-2 replied:
Thanks for spotting, it's curious that wifi still did work (probably the driver does not enforce truncation but defaults to what it expects the size to be).

DragonBluep replied:
If the size is wrong, you can only get about 3 dBm low tx power.

DragonBluep reviewed (commented):
I would suggest add explicit pinctrl config.

s-2 commented:
> I would suggest add explicit pinctrl config.

This would just include `wled_an` then I suppose?
Will add and rebuild, thanks

DragonBluep commented:
> > I would suggest add explicit pinctrl config.
> 
> This would just include then I suppose? Will add and rebuild, thanks`wled_an`

```
&state_default {
	gpio {
		groups = "gpio", "refclk", "wdt", "wled_an";
		function = "gpio";
	};
};
```
ref:
https://github.com/luowanhua/7628-docs/blob/main/MT7628_Datasheet.pdf
https://elixir.bootlin.com/linux/v6.6.32/source/Documentation/devicetree/bindings/pinctrl/mediatek,mt76x8-pinctrl.yaml

s-2 commented:
dropped `compatible = 'mac-address'`, fixed typo `macaddr@44`, fixed mtd eeprom size, added pinctrl, rebuilt and tested.

I wonder if it's acceptable to have the `&default-state` node in the .dtsi, since both devices indeed happen to use the same GPIO (Reset on Pin 38 / `wdt`) and LED Pins (11 / 37 / 44 i.e. `gpio / refclk / wled_an`). 
Personally I would prefer to have the LED pinctrl in the corresponding .dts, although it's redundant (and reset pin would be redundant, not sure if this should be separated further though).

DragonBluep commented:
> dropped `compatible = 'mac-address'`, fixed typo `macaddr@44`, fixed mtd eeprom size, added pinctrl, rebuilt and tested.
> 
> I wonder if it's acceptable to have the `&default-state` node in the .dtsi, since both devices indeed happen to use the same GPIO (Reset on Pin 38 / `wdt`) and LED Pins (11 / 37 / 44 i.e. `gpio / refclk / wled_an`). 

I think both are okay. Anyway the final device tree won't make any difference.

DragonBluep commented on target/linux/ramips/dts/mt7628an_7links_wlr-12xx.dtsi:
```diff
@@ -0,0 +1,134 @@
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		ieee80211-freq-limit = <5000000 6000000>;
+		nvmem-cells = <&eeprom_factory_8000>, <&macaddr_factory_8004>;
```
After correct the eeprom size,the mac address can be removed as it's already the default value (eeprom offset 4 bytes).

DragonBluep reviewed (approved):
Looks good to me. By the way, if you don't mind force push it again, please add some hardware information such as ROM/RAM size, and LED in the commit message.

s-2 commented:
> please add some hardware information such as ROM/RAM size, and LED in the commit message.

Is there anything that could be added here? RAM / ROM size and basic LED info are given in both commit messages.

DragonBluep commented:
> Is there anything that could be added here? RAM / ROM size and basic LED info are given in

Oh! My bad. I looked too fast to notice them. Every thing should be ready now.
